Kms provider doc

vineet-garg opened the pull request:
fixes [7399](https://github.com/kubernetes/website/issues/7399)

jhorwit2 commented on docs/tasks/administer-cluster/KMS-provider.md:
```diff
@@ -0,0 +1,172 @@
+  - resources:
+    - secrets
+    providers:
+    - kms:
```
indentation seems off.

vineet-garg replied:
fixed and tested

jhorwit2 commented on docs/tasks/administer-cluster/KMS-provider.md:
```diff
@@ -0,0 +1,172 @@
+    providers:
+    - identity: {}
+    - kms:
+      name : <display name for the KMS plugin>
```
indentation seems off.

vineet-garg replied:
fixed and tested

jhorwit2 commented on docs/tasks/administer-cluster/KMS-provider.md:
```diff
@@ -0,0 +1,172 @@
+
+The KMS encryption provider uses an envelope encryption scheme to encrypt data in etcd. The Key encryption keys (KEKs) are 
+stored and managed in a remote KMS. The KMS provider uses gRPC to communicate with a specific KMS 
+plugin. The KMS plugin, which is implemented as a gRPC server and deployed on the same host as the Kubernetes master, is responsible for all communication with the remote KMS.
```
What about HA masters?

vineet-garg replied:
@jhorwit2 
The grpc server should be deployed on all kubernetes masters hosts if there are more than 1 master deployed on multiple hosts. Do you think I should mention this explicitly?

jhorwit2 commented on docs/tasks/administer-cluster/KMS-provider.md:
```diff
@@ -0,0 +1,172 @@
+are stored and managed by the KMS plugin independently. The KMS plugin can encode the ciphertext with additional metada that may be required before sending it to the KMS for decryption.
+
+### Deploying the KMS plugin 
+Ensure that the KMS plugin runs on the same host as the Kubernetes master. Cluster administrators can use various options, such as taints and tolerations, to deploy the KMS plugin on the master host.
```
You'd want to use node selectors + tolerations for most installations. Tolerations won't force the scheduler to only pick masters; you'll need node selectors for that.

jhorwit2 replied:
@vineet-garg I'd just remove the second sentence since that's cluster dependent.

kksriram commented on docs/tasks/administer-cluster/KMS-provider.md:
```diff
@@ -70,7 +70,7 @@ All configuration data, including authentication credentials the KMS plugin uses
 are stored and managed by the KMS plugin independently. The KMS plugin can encode the ciphertext with additional metada that may be required before sending it to the KMS for decryption.
```
Typo. metadata.

vineet-garg replied:
corrected

kksriram commented on docs/tasks/administer-cluster/KMS-provider.md:
```diff
@@ -0,0 +1,172 @@
+
+* {% include task-tutorial-prereqs.md %}
+
+* Kubernetes version 1.7.0 or later is required
```
1.10.0 or later?

vineet-garg replied:
fixed to 1.10.0

mayakacz commented on docs/tasks/administer-cluster/encrypt-data.md:
```diff
@@ -78,7 +78,8 @@ Name | Encryption | Strength | Speed | Key Length | Other Considerations
 `secretbox` | XSalsa20 and Poly1305 | Strong | Faster | 32-byte | A newer standard and may not be considered acceptable in environments that require high levels of review. 
-`aesgcm` | AES-GCM with random nonce | Must be rotated every 200k writes | Fastest | 16, 24, or 32-byte | Is not recommended for use except when an automated key rotation scheme is implemented. 
+`aesgcm` | AES-GCM with random nonce | Must be rotated every 200k writes | Fastest | 16, 24, or 32-byte | Is not recommended for use except when an automated key rotation scheme is implemented.
+`KMS` | Data is encrypted using AES-CBC with PKCS#7 padding, DEKs are encrypted according to configuration in KMS | Strongest, new DEK is generated for each encryption | Fast | 32-bytes for data. Key length for DEKs is configured in KMS | [Configure the KMS provider](docs/tasks/administer-cluster/KMS-provider)
```
We need to define "DEK" before it's used here.

Maybe a line earlier saying something like, "An additional option is envelope encryption, where secrets are initially encrypted with the data encryption key (DEK), and the DEK is itself encrypted with a key encryption key (KEK). This simplifies key rotation and limits the throughput needed for the key management service."

mayakacz replied:
"new DEK is generated for each encryption" is correct, but an implementation detail, not a comment on "Strongest". Would include at the end, e.g., "Strongest | Fast | 32-byte | The recommended choice for using a third party tool for key management. Simplifies key rotation, with a new DEK generated for each encryption, and KEK rotation controlled by the user."

mayakacz commented on docs/tasks/administer-cluster/KMS-provider.md:
```diff
@@ -0,0 +1,172 @@
+
+{% capture steps %}
+
+The KMS encryption provider uses an envelope encryption scheme to encrypt data in etcd. The Key encryption keys (KEKs) are 
```
add details on DEK, e.g.,
"The KMS encryption provider uses an envelope encryption scheme to encrypt data in etcd, using a data encryption key (DEK). The DEKs are encrypted with a key encryption key (KEK), stored and managed in a remote KMS..."

mayakacz commented on docs/tasks/administer-cluster/KMS-provider.md:
```diff
@@ -0,0 +1,172 @@
+
+  * `name`: Display name of the KMS plugin.
+  * `endpoint`: Listen address of the gRPC server (KMS plugin). The endpoint is a UNIX domain socket connection.
+  * `cachesize`: Number of data encryption keys (DEKs) to be cached in the clear. DEKs are used in Envelope encryption. 
```
"Number of data encryption keys (DEKs) to be cached in the clear. When cached, DEKs can be used without another call to the KMS; whereas DEKs that are not cached require a call to the KMS to unwrap."

vineet-garg commented:
@mayakacz Thanks for the comments. The pull request to kubernetes/kubernetes-website is https://github.com/kubernetes/website/pull/7479
I have addressed your comments there.
